### PR TITLE
[NEAT-154] DMS Extension Validation

### PR DESCRIPTION
**Pull request by doctrino**
Core change is 100-150 lines of code.

### Validation
When setting `schema=extended` this change validates the following comparint user to ref rules:
* If `reshape` and change in container -> Error
* If `addition` and change in container or view -> Error.

### Additional Bugfixes
* `model_dump` did not include `reference` and `is_reference` of `dms_rules`.
* `NeatValidationError` now unpacks `MultiValueError` correctly in `from_pydantic_errors`

### CODE REVIEW

File: cognite/neat/rules/models/_rules/dms_architect_rules.py [hunk 1]
```diff
@@ -490,12 +490,117 @@ def referenced_views_and_containers_are_existing(self) -> "DMSRules":
             raise issues.MultiValueError(errors)
         return self
 
+    @model_validator(mode="after")
+    def validate_extension(self) -> "DMSRules":
+        if self.metadata.schema_ is not SchemaCompleteness.extended:
+            return self
+        if not self.reference:
```
Comment (nikokaoja) [inline cognite/neat/rules/models/_rules/dms_architect_rules.py:497]: nice

File: cognite/neat/rules/models/_rules/dms_architect_rules.py [hunk 2]
```diff
@@ -490,12 +490,117 @@ def referenced_views_and_containers_are_existing(self) -> "DMSRules":
...
+            raise ValueError("The schema is set to 'extended', but no reference rules are provided to validate against")
+        is_solution = self.metadata.space != self.reference.metadata.space
+        if is_solution:
+            return self
+        if self.metadata.extension is ExtensionCategory.rebuild:
```
Comment (nikokaoja) [inline cognite/neat/rules/models/_rules/dms_architect_rules.py:502]: simple and effective !

File: cognite/neat/rules/models/_rules/dms_architect_rules.py [hunk 3]
```diff
@@ -490,12 +490,117 @@ def referenced_views_and_containers_are_existing(self) -> "DMSRules":
...
+            # Everything is allowed
+            return self
+        # Is an extension of an existing model.
+        user_schema = self.as_schema()
+        ref_schema = self.reference.as_schema()
+        new_containers = {container.as_id(): container for container in user_schema.containers}
+        existing_containers = {container.as_id(): container for container in ref_schema.containers}
```
Comment (nikokaoja) [inline cognite/neat/rules/models/_rules/dms_architect_rules.py:509]: we should consider updating base structure for all our sheets (except metdata), as currently they do not provide mechanism for fast accesses via keys to specific "rows". Rules v1 , we have access through "key", which is a unique ID of a thing (e.g. property, or class). I often find myself doing the same as you do here (rebuild a data structure for easy access)

Comment (nikokaoja) [reply cognite/neat/rules/models/_rules/dms_architect_rules.py:509]: Current structure allows for repetition and redefinition

File: cognite/neat/rules/models/_rules/dms_architect_rules.py [hunk 4]
```diff
@@ -490,12 +490,117 @@ def referenced_views_and_containers_are_existing(self) -> "DMSRules":
...
+
+        errors: list[issues.NeatValidationError] = []
+        for container_id, container in new_containers.items():
+            existing_container = existing_containers.get(container_id)
```
Comment (nikokaoja) [inline cognite/neat/rules/models/_rules/dms_architect_rules.py:513]: should this be `.get(container_id, None)` ?

File: cognite/neat/rules/models/_rules/dms_architect_rules.py [hunk 5]
```diff
@@ -490,12 +490,117 @@ def referenced_views_and_containers_are_existing(self) -> "DMSRules":
...
+            # Everything is allowed
+            return self
+        # Is an extension of an existing model.
+        user_schema = self.as_schema()
+        ref_schema = self.reference.as_schema()
+        new_containers = {container.as_id(): container for container in user_schema.containers}
+        existing_containers = {container.as_id(): container for container in ref_schema.containers}
+
+        errors: list[issues.NeatValidationError] = []
+        for container_id, container in new_containers.items():
```
Comment (nikokaoja) [inline cognite/neat/rules/models/_rules/dms_architect_rules.py:512]: maybe turn your review of containers as dedicated method to be a bit readable.

File: cognite/neat/rules/models/_rules/dms_architect_rules.py [hunk 6]
```diff
@@ -490,12 +490,117 @@ def referenced_views_and_containers_are_existing(self) -> "DMSRules":
...
+        if self.metadata.extension is ExtensionCategory.reshape and errors:
+            raise issues.MultiValueError(errors)
+        elif self.metadata.extension is ExtensionCategory.reshape:
+            # Reshape allows changes to views
+            return self
+
+        new_views = {view.as_id(): view for view in user_schema.views}
+        existing_views = {view.as_id(): view for view in ref_schema.views}
+        for view_id, view in new_views.items():
+            existing_view = existing_views.get(view_id)
```
Comment (nikokaoja) [inline cognite/neat/rules/models/_rules/dms_architect_rules.py:539]: same as above, I think you might need `.get(view_id, None)`

File: cognite/neat/rules/models/_rules/dms_architect_rules.py [hunk 7]
```diff
@@ -490,12 +490,117 @@ def referenced_views_and_containers_are_existing(self) -> "DMSRules":
...
+                    changed_attributes=changed_attributes or None,
+                )
+            )
+
+        if errors:
+            raise issues.MultiValueError(errors)
+        return self
+
+    @staticmethod
+    def _changed_attributes_and_properties(
```
Comment (nikokaoja) [inline cognite/neat/rules/models/_rules/dms_architect_rules.py:559]: would `deepdiff` be helpful here?

File: cognite/neat/rules/models/_rules/dms_architect_rules.py [hunk 8]
```diff
@@ -490,12 +490,117 @@ def referenced_views_and_containers_are_existing(self) -> "DMSRules":
...
+        new_attributes = {key: value for key, value in new_dumped.items() if key != "properties"}
+        existing_attributes = {key: value for key, value in existing_dumped.items() if key != "properties"}
+        changed_attributes = [key for key in new_attributes if new_attributes[key] != existing_attributes.get(key)]
+        new_properties = new_dumped.get("properties", {})
+        existing_properties = existing_dumped.get("properties", {})
+        changed_properties = [prop for prop in new_properties if new_properties[prop] != existing_properties.get(prop)]
+        return changed_attributes, changed_properties
+
     @model_validator(mode="after")
     def validate_schema(self) -> "DMSRules":
```
Comment (nikokaoja) [inline cognite/neat/rules/models/_rules/dms_architect_rules.py:572]: I will copy/paste your approach for `InformationRules` this look way way more readable then what I implemented

**Review by nikokaoja — APPROVED** [review]

File: cognite/neat/rules/models/_rules/dms_architect_rules.py (hunk 4 above)
Comment (doctrino) [reply cognite/neat/rules/models/_rules/dms_architect_rules.py:513]: That is always the default.

**Pull request merged**